src/tds/ncbi_strerror.c

```diff
@@ -0,0 +1,351 @@
+ *
+ *			      PUBLIC DOMAIN NOTICE
+ *		 National Center for Biotechnology Information
+ *
+ *  This software/database is a "United States Government Work" under the
+ *  terms of the United States Copyright Act.  It was written as part of
+ *  the author's official duties as a United States Government employee and
+ *  thus cannot be copyrighted.	 This software/database is freely available
+ *  to the public for use. The National Library of Medicine and the U.S.
+ *  Government have not placed any restriction on its use or reproduction.
+ *
+ *  Although all reasonable efforts have been taken to ensure the accuracy
+ *  and reliability of the software and data, the NLM and the U.S.
+ *  Government do not and cannot warrant the performance or results that
+ *  may be obtained by using this software or data. The NLM and the U.S.
+ *  Government disclaim all warranties, express or implied, including
+ *  warranties of performance, merchantability or fitness for any particular
+ *  purpose.
+ *
+ *  Please cite the author in any work or product based on this material.
+ *
+ * ===========================================================================
+ *
+ * Authors:  Pavel Ivanov, Anton Lavrentiev, Denis Vakatov
+ *
+ * File Description:
+ *   errno->text conversion helper
+ */
+
+
+#ifdef NCBI_INCLUDE_STRERROR_C
+
+
+#  ifdef _FREETDS_LIBRARY_SOURCE
+#    define s_StrError	      s_StrErrorInternal
+#  endif /*_FREETDS_LIBRARY_SOURCE*/
+
+
+#  if defined(NCBI_OS_MSWIN)  &&  defined(_UNICODE)
+
+static const char* s_WinStrdup(const char* str)
+{
+    size_t n = strlen(str);
+    char*  s = (char*) LocalAlloc(LMEM_FIXED, ++n * sizeof(*s));
+    return s ? (const char*) memcpy(s, str, n) : 0;
+}
+#    define	   ERR_STRDUP(s)	  s_WinStrdup(s)
+
+extern const char* UTIL_TcharToUtf8(const TCHAR* str)
+{
+    char* s = NULL;
+    if (str) {
+	/* Note "-1" means to consume all input including the trailing NUL */
+	int n = WideCharToMultiByte(CP_UTF8, 0, str, -1, NULL, 0, NULL, NULL);
+	if (n > 0) {
+	    s = (char*) LocalAlloc(LMEM_FIXED, n * sizeof(*s));
+	    if (s)
+		WideCharToMultiByte(CP_UTF8, 0, str, -1, s,    n, NULL, NULL);
+	}
+    }
+    return s;
+}
+
+#    ifndef	   UTIL_ReleaseBuffer
+#      define	   UTIL_ReleaseBuffer(x)   UTIL_ReleaseBufferOnHeap(x)
+#    endif
+
+#  else /*NCBI_OS_MSWIN && _UNICODE*/
+
+#    define	   ERR_STRDUP(s)	   strdup(s)
+
+#    ifdef	   UTIL_TcharToUtf8
+#      undef	   UTIL_TcharToUtf8
+#    endif
+#    define	   UTIL_TcharToUtf8(x)	   strdup(x)
+
+#    ifdef	   UTIL_ReleaseBuffer
+#      undef	   UTIL_ReleaseBuffer
+#    endif
+#    define	   UTIL_ReleaseBuffer(x)   free((void*)(x))
+
+#  endif /*NCBI_OS_MSWIN && _UNICODE*/
+
+
+#  ifdef NCBI_OS_MSWIN
+
+extern void	   UTIL_ReleaseBufferOnHeap(const void* ptr)
+{
+    if (ptr)
+	LocalFree((HLOCAL) ptr);
+}
+
+#  endif /*NCBI_OS_MSWIN*/
+
+
+static const char* s_StrErrorInternal(int error)
```

Comment:
Why not reusing `tds_prwsaerror` and `tds_prwsaerror_free` ?

Reply:
Good idea, they may well suffice.

Reply:
Nice, commit applied (slightly changed using some already existing macros to reduce conditional code).